sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java

```diff
@@ -0,0 +1,747 @@
+import org.apache.beam.sdk.coders.CustomCoder;
+import org.apache.beam.sdk.coders.KvCoder;
+import org.apache.beam.sdk.coders.VarLongCoder;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Combine.CombineFn;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+/**
+ * {@code PTransform}s for computing the most frequent elements of a stream.
+ *
+ * <p>This class uses the Stream Summary sketch which uses a HashMap to keep track of
+ * a certain number of elements considered as the most frequent elements in a stream.
+ *
+ * <h2>Brief overview of the sketch</h2>
+ *
+ * <p>The Stream Summary uses a doubly linked-list of buckets in order to order elements
+ * by frequency. Each bucket is associated with a list of Counters, each of them storing
+ * information about one of the tracked elements : value, count and maximum potential error.
+ * <br>
+ *   An element is guaranteed to be in the top K most frequent if its guaranteed number
+ *   of hits, i.e. the {@code count minus error}, is greater than the count of the element at
+ *   the position k+1.
+ * <br>
+ *   <b>WARNING</b>: Information about the error is lost during the parallelization,
+ *   so this value may be underestimated in most cases, that is why the error is
+ *   not output by default. This issue will be solved in the next version of the extension.
+ *
+ * <h2>References</h2>
+ *
+ * <p>This class uses the Space-Saving algorithm, introduced in the paper available <a href=
+ * "https://pdfs.semanticscholar.org/72f1/5aba2e67b1cc9cd1fb12c99e101c3c1aae3b.pdf">here</a>.
+ * <br>The implementation of the Stream Summary sketch comes from
+ * <a href="https://github.com/addthis/stream-lib">Addthis' library Stream-lib</a>:
+ *
+ * <h2>Parameters</h2>
+ *
+ * <p>Two parameters can be tuned:
+ *
+ * <ul>
+ *   <li>The top {@code k} most frequent elements you want to output from the stream.<br>
+ *     By default, {@code k} is set to {@code 100}.
+ *   <li>The {@code capacity} of the Stream Summary sketch. It corresponds to
+ *   the maximum number of elements that can be tracked at once.
+ *   When the maximum {@code capacity} is reached, the least frequent element
+ *   will be replaced by the next untracked element.<br>
+ *     By default, the {@code capacity} is set to {@code 10000} but this value depends
+ *     highly on the use case and more specifically on the order of incoming elements.
+ * </ul>
+ *
+ *  <p><b>WARNING</b>: Keep in mind that he {@code capacity} should always be
+ *  significantly higher than {@code k}. One must ask himself the following question:<br>
+ *    What is the lowest rank {@code x} such that any element arriving at a rank {@code y > x}
```

Comment:
What does "arriving at a rank y" mean, considering lack of ordering?

Reply:
What I want to say is :
What is the minimal number of different elements N such that in the context of the time window of a processing, every new element will have a very low probability to be considered as a frequent element at the end of the time window.

Is it more clear from a beam context?

Reply:
Beam does not have the concept of arrival order of elements, period. A runner is technically allowed to accumulate stuff and then call your DoFn's on it in random order; and if you're using one of the runners or runner modes primarily designed for batch use cases, this is pretty much what it will be like.

So, in other words, we can only give guarantees that can be justified by the Beam model itself, rather than by assumptions about run-time execution.

Reply:
Ok I understand.